js/components/select.tsx

```diff
@@ -75,14 +76,8 @@ class Select extends React.Component<IItemsComponentProps, IDropDownState> {
             }
 
             if (state && state.objectData) {
-
-                const selectedOptions = state.objectData.filter(
-                    item => options.find(option => option.value.internalId === item.internalId),
-                );
-
-                if (selectedOptions.length === 0) {
-                    options = (this.getOptions(state.objectData) || []).concat(options);
-                }
+                // Replace 'selected' item in the option list. Match on `externalId` because the internalId is volatile
+                options = unionWith(eqBy(path(['value', 'externalId'])), this.getOptions(state.objectData), options);
```

Comment:
There is a Ramda function called eqProps that could cut down using eqBy & path.

Also, equivalence by externalId would not work in the case where there are multiple items added whilst offline.
Items added whilst offline do not have externalIds, so only the first offline-only item would be represented after the union:
[OnlineRamdaTester](https://ramdajs.com/repl/#?const%20list1%20%3D%20%5B%7Bid%3A%20%22test%22%2Cname%3A%20%221%22%7D%2C%7Bid%3A%20null%2Cname%3A%20%222%22%7D%5D%3B%0Aconst%20list2%20%3D%20%5B%7Bid%3A%20null%2Cname%3A%20%225%22%7D%2C%7Bid%3A%20null%2Cname%3A%20%223%22%7D%2C%7Bid%3A%20%22be%22%2Cname%3A%20%224%22%7D%5D%3B%0Aconsole.log%28unionWith%28eqBy%28path%28%5B%27id%27%5D%29%29%2Clist1%2Clist2%29%29%3B%0Aconsole.log%28unionWith%28eqProps%28%27id%27%29%2Clist1%2Clist2%29%29%3B)
In this test you can see the result has removed the multiple values whose id are null, but these could be all valid and different, just that they are offline-only and haven't got an externalId (I called the prop id).

Reply:
Unfortunately eqProps doesn't handle nested properties, hence the eqBy & path

I wasted a couple of hours of my life in the ramda playground trying to match on internal || external IDs but no joy. It got messy with nested OR predicates and superfluous checks.

The new fix does not use ramda but with the added bonus that the order of items in the select  is preserved. ramda did not guarantee the order so not very useful.

Now checks externalId and falls back to internalId to handle the offline case.

Reply:
Ah, I didn't realize.
The new fix looks good.